### ISSUE TITLE
Unable to decode storage.staking.validators

### ISSUE THREAD

**Issue by wulforr** [polkadot-js/api#1442]
I am trying to access intentions for Kusama chain. The code that i am using is
```let provider = new WsProvider("wss://kusama-rpc.polkadot.io");
  const apinew = await ApiPromise.create({ provider });
  
  const intentions = await apinew.query.staking.validators()
  console.log(JSON.parse(JSON.stringify(intentions)))
  ```
This code is giving me an error:
```
Unable to decode storage staking.validators: createType((ValidatorPrefs, Linkage<AccountId>)):: U8a: failed on 'Type':: U8a: failed on 'Compact':: invalid array length

RPC-CORE: subscribeStorage (keys: Vec<StorageKey>): StorageChangeSet:: createType((ValidatorPrefs, Linkage<AccountId>)):: U8a: failed on 'Type':: U8a: failed on 'Compact':: invalid array length
```

**Comment by jacogr** [comment]
Weird, which version of the API?

**Comment by wulforr** [comment]
I am using 0.93.1
Other information about Kusama network is fetching perfectly.

**Comment by jacogr** [comment]
Ok, then that is even weirder, since we def. didn't see it in use, for instance on the apps UI. 

Also just checked on the apps UI in the JS console and it is all happy executing that one - it is a bit ahead of 0.93.1 atm since it always runs the beta versions of the API.

If you feel a bit adventurous, could you try the `0.94.0-beta.10` version? (It actually only has a couple of minor fixes from 0.93.1, i.e. nothing "breaking" - trying to establish some sort of way to get to the bottom of this).

I'm assuming your provider is a local node? Or one of the "public" nodes?

**Comment by wulforr** [comment]
I tried the beta version but the problem still persists.
Also, I could not understand what you are trying to say in the last line. Could you please elaborate on how to check it?

**Comment by jacogr** [comment]
So really trying to understand what this line refers to -

`const apinew = await ApiPromise.create({ provider });`

Which provider endpoint are you using? Since not seeing the same at all with any of my tests on Kusama, hence the question.

**Comment by wulforr** [comment]
```
 let provider = new WsProvider("wss://kusama-rpc.polkadot.io");
```
I am using WsProvider

**Comment by jacogr** [comment]
I've added a E2E test with exactly that here - https://github.com/polkadot-js/api/pull/1448/files#diff-3ea27db0e2275f06b36fe15c3e631d3aR1

Which yields no errors as well, same as the apps UI.

```
Kusama
    ✓ retrieves the list of validators (4233ms)

  console.error packages/api/test/e2e/api/promise-kusama.spec.ts:20
    [
      [
        'CrDvXaVGt8HFWGyKGUGwKW1a3AFvfmY1y46NVcgtjcgoa7g',
        'DMyeifp3hWAoZx7jC3QkthvSeBmr7rzJSuSmgLSY74xAUD4',
        ... removed a bundle for readability
        'ET9SkhNZhY7KT474vkCEJtAjbgJdaqAGW4beeeUJyDQ3SnA',
        'DGUYMCQSdPJsioaQW6ecAAGbsr1HWHbK9cVng6gCBtXG4Ni'
      ],
      [
        { validatorPayment: 1000000000000 },
        { validatorPayment: 100000000000 },
       ... remove a bundle for readability
        { validatorPayment: 0 },
        { validatorPayment: 0 }
      ]
    ]

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        7.496s, estimated 13s
Ran all test suites matching /packages\/api\/test\/e2e\/api\/promise-kusama.spec.ts/i.
✨  Done in 8.68s.
```

Do you have a mismatch in your types and API (i.e. if you have both packages, are they the same version?)

**Comment by wulforr** [comment]
```
import {WsProvider, ApiPromise} from '@polkadot/api'


export default class App extends Component {

  componentDidMount() {
    this.start()
    this.kusamastart()
  }
  
  async start() {
  let provider = new WsProvider("wss://poc3-rpc.polkadot.io");
  const api = await ApiPromise.create({ provider });
  
  const intentions = await api.query.staking.validators()
  console.log(JSON.parse(JSON.stringify(intentions)))

  }

  async kusamastart(){
    let provider = new WsProvider("wss://kusama-rpc.polkadot.io");
    const apinew = await ApiPromise.create({ provider });
  
    const intentions = await apinew.query.staking.validators()
    console.log(JSON.parse(JSON.stringify(intentions)))
    })
  
  }
  

```

Using both Kusama and Alexander is causing an error.
Kusama returns the error
while Alexander returns an array of validators but they are not same as validators and intentions in 
[Polkadot.js.org](https://polkadot.js.org/apps/#/staking)

**Comment by jacogr** [comment]
Ok, please post your `package.json` 

Yes, there is a difference between what Kusama (Substrate 2.x) and Alexander (Substrate 1.x, >6 months old) return.

**Comment by wulforr** [comment]
```
{
  "name": "practice",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@polkadot/api": "^0.93.1",
    "konva": "^4.0.12",
    "react": "^16.8.4",
    "react-dom": "^16.8.4",
    "react-konva": "^16.9.0-1",
    "react-scripts": "^2.1.8"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": "react-app"
  },
  "browserslist": [
    ">0.2%",
    "not dead",
    "not ie <= 11",
    "not op_mini all"
  ]
}
```
This is my package.json

**Comment by jacogr** [comment]
There is nothing strange in there at all. 

Looking back at your code - The type registry is global and there is a different `ValidatorPrefs` for Alexander. (So if you try and start both with a global registry, one of the two will fail since it is injected globally on the type of chain.) Only start one of the two.

**Comment by wulforr** [comment]
I want to access information regarding both Alexander and Kusama.
Could you guide me how can i achieve that?

**Comment by jacogr** [comment]
Currently, because of the global type registry, it is not possible to do within the same process/app where there are mismatching types. (as you have with Substrate 1.x and Substrate 2.x systems) Here is a snapshot of all the types that are being registered to allow operation with the srml - https://github.com/polkadot-js/api/blob/master/packages/types/src/interfaceRegistry.ts

As soon as the types are moved to being exposed via metadata and injected into the API (i.e. the JS API does not have a duplication of all types as defined in Rust), the intent it do attach this to the API instance itself instead of globally.

Sadly, this is not something that will be addressed before we have that capability, since it will be throw-away work. As for ETA, well, it is being tested as a localized registry on the contracts side, but as of now no work has been started in regards to this as of yet for the metadata and srml module types.

TL;DR You can have multiple instances where the base Substrate types do not conflict, but not with conflicts. (ValidatorPrefs, BlockNumbers, Account Nonces are all different between 1.x and 2.x default chains)
